File: Data Structures & Algorithms/combination-target-sum/submission-9.py
# [2, 5, 6, 9]
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        self.ret = []

        def backtrack(index, subset, subsetSum):
            if subsetSum > target:
                return
            if subsetSum == target:
                for i in self.ret:
                    if (i == subset):
                        return
                self.ret.append(subset[:])
                return
            if (index != 0):
                subset.append(nums[index-1])
                backtrack(index, subset, subsetSum + nums[index-1])
                subset.pop()
            if (index == len(nums)):
                return

            subset.append(nums[index])
            backtrack(index + 1, subset, subsetSum + nums[index])
            subset.pop()

            backtrack(index + 1, subset, subsetSum)
            return
        backtrack(0, [], 0)
        return self.ret

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-9.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        int L = 0;
        int R = nums.size() - 1;
        while (L < R) {
            int m = (L + R) / 2;
            cout << m << endl;
            if (R - L == 1) {
                return min(nums[R], nums[L]);
            }
            if (nums[R] < nums[L] and nums[m] > nums[L]) {
                L = m + 1;
            } else {
                R = m;
            }


        }
        return nums[L];
    }
};
